src/common/type/stringId.c

```diff
@@ -0,0 +1,375 @@
+
+                case 4:
+                    result |= (uint64_t)map[(uint8_t)buffer[3]] << 22;
+
+                case 3:
+                    result |= (uint64_t)map[(uint8_t)buffer[2]] << 16;
+
+                case 2:
+                    result |= (uint64_t)map[(uint8_t)buffer[1]] << 10;
+
+                case 1:
+                    result |= (uint64_t)map[(uint8_t)buffer[0]] << 4;
+            }
+
+            FUNCTION_TEST_RETURN(result);
+        }
+    }
+}
+
+/**********************************************************************************************************************************/
+size_t
+strIdToZN(StringId strId, char *const buffer)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(STRING_ID, strId);
+        FUNCTION_TEST_PARAM(VOID, buffer);
+    FUNCTION_TEST_END();
+
+    ASSERT(strId != 0);
+    ASSERT(buffer != NULL);
+
+    // Is the StringId a prefix of a longer string?
+    bool prefix = strId & STRING_ID_PREFIX;
+
+    // Extract bits used to encode the characters
+    StringIdBit bit = (StringIdBit)(strId & STRING_ID_BIT_MASK);
+
+    // Remove header to get the encoded characters
+    strId >>= STRING_ID_HEADER_SIZE;
+
+    // Decoding type
+    switch (bit)
+    {
+        // 5-bit decoding
+        case stringIdBit5:
+        {
+            // Map to convert encoding to characters
+            const char map[32] = "!abcdefghijklmnopqrstuvwxyz-1234";
+
+            // Macro to decode all but the last character
+            #define STR5ID_TO_ZN_IDX(idx)                                                                                          \
+                buffer[idx] = map[strId & 0x1F];                                                                                   \
+                strId >>= 5;                                                                                                       \
+                                                                                                                                   \
+                if (strId == 0)                                                                                                    \
+                    FUNCTION_TEST_RETURN(idx + 1)
+
+            // Char 1-11
+            STR5ID_TO_ZN_IDX(0);
+            STR5ID_TO_ZN_IDX(1);
+            STR5ID_TO_ZN_IDX(2);
+            STR5ID_TO_ZN_IDX(3);
+            STR5ID_TO_ZN_IDX(4);
+            STR5ID_TO_ZN_IDX(5);
+            STR5ID_TO_ZN_IDX(6);
+            STR5ID_TO_ZN_IDX(7);
+            STR5ID_TO_ZN_IDX(8);
+            STR5ID_TO_ZN_IDX(9);
+            STR5ID_TO_ZN_IDX(10);
+
+            // Char 12
+            buffer[11] = map[strId & 0x1F];
+            ASSERT(strId >> 5 == 0);
+
+            // If prefix flag is set then append +
+            if (prefix)
+            {
+                buffer[12] = '+';
+                FUNCTION_TEST_RETURN(13);
+            }
+
+            FUNCTION_TEST_RETURN(12);
```

Comment:
So this is STRID5_MAX - 1?

Reply:
Yes, but then one wonders whether we should do STRID5_MAX - 2, STRID5_MAX - 3, etc.

It's a constant, but it cannot ever be changed without breaking everything. The only reason it exists is to help the caller properly size a buffer so not sure it make sense to propagate it through the code.